runner/src/internal.rs

```diff
@@ -223,66 +247,153 @@ pub fn run_step(context: &Context, step: &Step) -> HashSet<StatusResultValue> {
 }
 
 impl Cmd {
-    pub fn new<I, S>(executable: &str, args: I) -> Self
+    pub fn new<I, S>(executable: &str, args: I) -> Box<Self>
     where
         I: IntoIterator<Item = S>,
         S: AsRef<str>,
     {
-        Cmd {
+        Box::new(Cmd {
             executable: executable.to_string(),
             args: args.into_iter().map(|s| s.as_ref().to_string()).collect(),
-        }
+            env: HashMap::new(),
+        })
+    }
+
+    pub fn new_with_env<I, S>(executable: &str, args: I, env: &HashMap<String, String>) -> Box<Self>
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<str>,
+    {
+        Box::new(Cmd {
+            executable: executable.to_string(),
+            args: args.into_iter().map(|s| s.as_ref().to_string()).collect(),
+            env: env.clone(),
+        })
+    }
+}
+
+impl Runnable for Cmd {
+    fn description(&self) -> String {
+        format!("{} {}", self.executable, self.args.join(" "))
     }
     /// Run the provided command, printing a status message with the current prefix.
     /// TODO(#396): Return one of three results: pass, fail, or internal error (e.g. if the binary
     /// to run was not found).
-    fn run(&self, opt: &Opt) -> Running {
-        std::io::stderr().flush().expect("could not flush stderr");
-        let mut cmd = std::process::Command::new(&self.executable);
+    fn run(self: Box<Self>, opt: &Opt) -> Box<dyn Running> {
+        let mut cmd = tokio::process::Command::new(&self.executable);
         cmd.args(&self.args);
-        let command_string = format!("{:?}", cmd);
+
+        // Clear the parent environment. Only the variables explicitly passed below are going to be
+        // available to the command, to avoid accidentally depending on extraneous ones.
+        cmd.env_clear();
+
+        // General variables.
+        cmd.env("HOME", std::env::var("HOME").unwrap());
+        cmd.env("PATH", std::env::var("PATH").unwrap());
+        if let Ok(v) = std::env::var("USER") {
+            cmd.env("USER", v);
+        }
+
+        // Python variables.
+        if let Ok(v) = std::env::var("PYTHONPATH") {
+            cmd.env("PYTHONPATH", v);
+        }
+
+        // Rust compilation variables.
+        if let Ok(v) = std::env::var("RUSTUP_HOME") {
+            cmd.env("RUSTUP_HOME", v);
+        }
+        if let Ok(v) = std::env::var("CARGO_HOME") {
+            cmd.env("CARGO_HOME", v);
+        }
+
+        // Rust runtime variables.
+        cmd.env(
+            "RUST_LOG",
+            std::env::var("RUST_LOG").unwrap_or_else(|_| "info".to_string()),
+        );
+        cmd.env("RUST_BACKTRACE", "1");
+
+        // Emscripten variables.
+        if let Ok(v) = std::env::var("EMSDK") {
+            cmd.env("EMSDK", v);
+        }
+        if let Ok(v) = std::env::var("EM_CACHE") {
+            cmd.env("EM_CACHE", v);
+        }
+        if let Ok(v) = std::env::var("EM_CONFIG") {
+            cmd.env("EM_CONFIG", v);
+        }
+
```

Comment:
What does this do? Is it reading the env variables?

Reply:
Looking at L348 it seems that's what it does, which I suppose answers it. Seems like a verbose way of writing it?